### PR TITLE
Shortcut implementation

### PR DESCRIPTION
**Pull request by omarabid**
Shortcut implementation for ExpressCheckout and ExpressCheckout for Digital Goods

### CODE REVIEW

File: wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php [hunk 1]
```diff
@@ -7,37 +7,178 @@
...
+		$purchase_log->set( array(
+			'user_ID'        => wpsc_get_current_customer_id(),
+			'date'           => time(),
+			'plugin_version' => WPSC_VERSION,
+			'statusno'       => '0',
+			'sessionid'      => $sessionid,
+        ) );
+
+        // Set a Transient for EC Shortcut
+        set_transient( 'ecs-' . $sessionid, true, 60 * 60 );
```
Comment (JeffPyeBrook) [inline wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php]: Is using a transient the best thing here?  Transients can be deleted or unavailable at any time.

File: wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php [hunk 2]
```diff
@@ -7,37 +7,178 @@
...
+            return;
+        }
+        $payment_gateway = $_GET['payment_gateway'];
+        
+		global $wpsc_cart;
+		//	Create a new PurchaseLog Object 
+		$purchase_log = new WPSC_Purchase_Log();	
+
+		// Create a Sessionid
+		$sessionid = ( mt_rand( 100, 999 ) . time() );
```
Comment (JeffPyeBrook) [inline wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php]: This doesn't seem very unique? If you need a unique number that applies only to a specific shopper you could use the customer id

File: wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php (hunk 1 above)
Comment (JeffPyeBrook) [inline wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php]: If you need to save a value that applies to only the current shopper session consider using the customer meta functions.

Comment (omarabid) [inline wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php]: @JeffPyeBrook I don't want these meta to be persistent after being used, that's why I'm using transients here.

File: wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php (hunk 2 above)
Comment (omarabid) [inline wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php]: That line (and that part of code) were taken from the Checkout code for t.e. 2.0. So if I'm changing it here, it should be changed too for t.e. 2.0 for the same thing ( https://github.com/wp-e-commerce/WP-e-Commerce/blob/master/wpsc-components/theme-engine-v2/mvc/controllers/checkout.php#L90 ) and have it documented or maybe create a function for it.

@JustinSainton - @JeffPyeBrook What are your suggestions here?

File: wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php (hunk 1 above)
Comment (JustinSainton) [inline wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php]: @omarabid You can't count on [transients _ever_ existing](http://journal.ryanmccue.info/296/youre-using-transients-wrong/).  If you don't want to have the meta data be persistent, that's fine, just remove them after a successful transaction has been completed.

File: wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php (hunk 2 above)
Comment (JustinSainton) [inline wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php]: I don't think the customer ID makes sense as the unique identifier here - as I assume you'd want it globally unique, across all of their purchases?  The unique identifier here is unique enough for most use cases, but not all.  

In theory, if you had 1,000 transactions occur per second, you'd be running into dupes...not really a huge issue, but one we may have to deal with some day.

All that to say, we could probably use the customer ID and the current unique session ID together for a totally unique ID.

Comment (JeffPyeBrook) [inline wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php]: What i was suggesting was using the value of the customer cookie, it is the highest entropy thing we have going and if it has a collision there are other issues that will be much more serious

File: wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php [hunk 3]
```diff
@@ -206,6 +349,136 @@ public function callback_ipn() {
 		exit;
 	}
 
+    /**
+	 * Pull and Record PayPal Details
+	 *
+	 * @return void
+	 */
+	public function pull_paypal_details() {
```
Comment (JustinSainton) [inline wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php]: Review some of the notes here from Scrutinizer...this one method pulls down the entire codebase from 7.64 to 7.61.

https://scrutinizer-ci.com/g/wp-e-commerce/WP-e-Commerce/inspections/ff58628f-69ef-4320-9d68-fb016dc36377/code-structure/operation/%2Bglobal%5CWPSC_Payment_Gateway_Paypal_Express_Checkout%3A%3Apull_paypal_details

**Comment by JustinSainton** [comment]
Closing in favor of #1822.

**Pull request merged**